packages/component/src/providers/CustomElements/customElements/CodeBlock.ts

```diff
@@ -0,0 +1,240 @@
+  }
+
+  get options() {
+    const { theme } = this;
+    return theme
+      ? {
+          theme
+        }
+      : undefined;
+  }
+
+  attributeChangedCallback() {
+    this.highlightedCodeFragment = null;
+    this.update();
+  }
+
+  connectedCallback() {
+    this.connected = true;
+    this.update();
+  }
+
+  disconnectedCallback() {
+    this.connected = false;
+  }
+
+  highlight(...args: Parameters<HighlightCodeFn>) {
+    return defaultHighlightCode(...args);
+  }
+
+  update() {
+    if (!this.connected) {
+      return;
+    }
+
+    const { code, language, options, ownerDocument: document } = this;
+
+    if (code && !this.highlightedCodeFragment) {
+      const highlightCodeFragment = this.constructHighlightedCode(code, language, options);
+      this.highlightedCodeFragment = highlightCodeFragment;
+      const body = highlightCodeFragment.querySelector('pre');
+      body?.classList.add('webchat__code-block__body');
+      options.theme && body?.classList.add(options.theme);
+    }
+
+    if (!this.#originalFragment) {
+      this.#originalFragment = document.createDocumentFragment();
+      this.#originalFragment.replaceChildren(...this.children);
+    }
+
+    const sourceFragment = this.highlightedCodeFragment ? this.highlightedCodeFragment : this.#originalFragment;
+    const fragment = sourceFragment?.cloneNode(true);
+
+    this.copyButtonElement ??= this.constructCopyButton();
+    this.copyButtonElement && fragment.insertBefore(this.copyButtonElement, fragment.firstChild);
+
+    this.replaceChildren(fragment);
+  }
+
+  constructCopyButton() {
+    return null;
+  }
+
+  constructHighlightedCode(...args: Parameters<HighlightCodeFn>) {
+    const result = this.highlight(...args);
+    return result instanceof DocumentFragment ? result : parseDocumentFragmentFromString(result);
+  }
+}
+
+const createReactCodeBlockClass = ({
+  codeBlockRef,
+  copyButtonRef,
+  trackCodeBlockRefChanges,
+  trackCopyButtonRefChanges
+}: {
+  codeBlockRef: RefObject<{
+    className: string;
+    highlightCode: HighlightCodeFn;
+    theme: StyleOptions['codeBlockTheme'];
+  }>;
+  copyButtonRef: RefObject<{
+    altCopied: string;
+    altCopy: string;
+    className: string;
+    tagName: string;
+  }>;
+  trackCodeBlockRefChanges: OnTrackFn;
+  trackCopyButtonRefChanges: OnTrackFn;
+}) =>
+  class ReactCodeBlock extends CodeBlock {
+    static observedAttributes = CodeBlock.observedAttributes;
+
+    get options() {
+      // Prefer options from element attributes over ones passed from the context
+      return {
+        theme: codeBlockRef.current.theme,
+        ...super.options
+      };
+    }
+
+    #controller: AbortController;
+    #prevClassName: string;
+
+    connectedCallback() {
+      super.connectedCallback();
+
+      if (!this.#controller) {
+        this.#controller = new AbortController();
+
+        trackCodeBlockRefChanges(() => {
+          this.highlightedCodeFragment = null;
```

Comment:
Maybe use `undefined`?

Reply:
What about null for object references and undefined for everything else? :rofl: